cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/Filters/FilterCollection/FilterShowlistInput/index.tsx

```diff
@@ -0,0 +1,191 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import IconButton from '@material-ui/core/IconButton';
+import withStyles, { StyleRules, WithStyles } from '@material-ui/core/styles/withStyles';
+import AddIcon from '@material-ui/icons/Add';
+import DeleteIcon from '@material-ui/icons/Delete';
+import classnames from 'classnames';
+import { IPropertyShowConfig } from 'components/ConfigurationGroup/types';
+import Heading, { HeadingTypes } from 'components/Heading';
+import If from 'components/If';
+import { SHOW_TYPE_VALUES } from 'components/PluginJSONCreator/constants';
+import PluginInput from 'components/PluginJSONCreator/Create/Content/PluginInput';
+import { ICreateContext, IWidgetInfo } from 'components/PluginJSONCreator/CreateContextConnect';
+import isNil from 'lodash/isNil';
+import * as React from 'react';
+import uuidV4 from 'uuid/v4';
+
+const styles = (): StyleRules => {
+  return {
+    showConfigCollection: {
+      display: 'grid',
+      gridAutoFlow: 'column',
+      width: '100%',
+    },
+    showConfigInput: {
+      gridRow: '1',
+      width: '100%',
+    },
+    showConfigNameInput: {
+      gridColumnStart: '1',
+      gridColumnEnd: '8',
+    },
+    showConfigTypeInput: {
+      gridColumnStart: '9',
+      gridColumnEnd: '15',
+    },
+    showAddDeleteButtonInput: {
+      gridColumnStart: '19',
+      gridColumnEnd: '20',
+    },
+    showAddDeleteButton: {
+      float: 'right',
+    },
+  };
+};
+
+interface IFilterShowlistInputProps extends WithStyles<typeof styles>, ICreateContext {
+  filterID: string;
+}
+
+const FilterShowlistInputView: React.FC<IFilterShowlistInputProps> = ({
+  classes,
+  filterID,
+  filterToShowList,
+  setFilterToShowList,
+  showToInfo,
+  setShowToInfo,
+  widgetInfo,
+}) => {
+  const allWidgetNames = widgetInfo
+    ? Object.values(widgetInfo)
+        .map((info: IWidgetInfo) => info.name)
+        .filter((widgetName) => !isNil(widgetName) && !isNil(widgetName))
+    : [];
+
+  function setShowProperty(showID: string, property: string) {
+    return (val) => {
+      setShowToInfo((prevObjs) => ({
+        ...prevObjs,
+        [showID]: { ...prevObjs[showID], [property]: val },
+      }));
+    };
+  }
+
+  function addShowToFilter(filterObjID: string, index?: number) {
+    const newShowID = 'Show_' + uuidV4();
+
+    setShowToInfo({
+      ...showToInfo,
+      [newShowID]: {
+        name: '',
+      } as IPropertyShowConfig,
+    });
+
+    if (index === undefined) {
+      setFilterToShowList({
+        ...filterToShowList,
+        [filterObjID]: [...filterToShowList[filterObjID], newShowID],
+      });
+    } else {
+      const showList = filterToShowList[filterObjID];
+
+      if (showList.length === 0) {
+        showList.splice(0, 0, newShowID);
+      } else {
+        showList.splice(index + 1, 0, newShowID);
```

Comment:
you might have change this in immutableJS PR, but just to make sure - this is mutating the original array, we should clone the array first

Reply:
This is resolved through immutable PR #12251  and refactor PR #12310 
https://github.com/cdapio/cdap/blob/882eb810d3291ed6b39efaf6394a932584abb867/cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/FilterPage/FilterPanel/FilterShowlistInput/index.tsx#L43